### PR TITLE
Add explosion/cython-blis

### PR DESCRIPTION
**Pull request by honnibal**
@conda-forge/help-python-c

This probably isn't quite right yet -- I expect a build failure at least on Windows.

### CODE REVIEW

**Comment by chrisburr** [comment]
I've only glanced at this, but it looks like the recipe is detecting the compilers incorrectly. The conda compiler binaries are prefixed with something like `x86_64-linux-gcc`. To get the compiler correctly you should use the `$CC`/`$CXX`/`$LD` environment variables.

**Comment by honnibal** [comment]
Thanks! It first checks an environment variable `BLIS_COMPILER`, so I think we'll be able to do something here.

**Comment by honnibal** [comment]
Hmm. What's the best way to set an environment variable? I tried this:

```
build:
    number: 0
    script: "BLIS_COMPILER=$CC {{ PYTHON }} -m pip install . --no-deps -vv"
```

But it doesn't seem to have worked. Do I need to make `build.sh` and `build.bat` files?

File: recipes/blis/build.sh [hunk 1]
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env bash
+
+export BLIS_COMPILER="$CC"
+{{ PYTHON }} -m pip install . --no-deps -vv
```
Comment (chrisburr) [inline recipes/blis/build.sh]: This won't work as `{{ PYTHON }}` is a jinja2 variable which will only be expanded in the recipe.
```suggestion
"$PYTHON" -m pip install . --no-deps -vv
```

**Comment by chrisburr** [comment]
You can also use:
```yaml
build:
    number: 0
    script:
      - export BLIS_COMPILER=$CC  # [unix]
      - whatever windows needs here...  # [win]
      - {{ PYTHON }} -m pip install . --no-deps -vv"
```
but I would use `build.sh` and `bld.bat` scripts while debugging and then decide if they're trivial enough to justify moving into the `meta.yaml` file.

**Comment by honnibal** [comment]
Ah, cool. I wasn't sure whether that would work...I've used other yaml build systems where the lines end up sort of independent.

**Comment by isuruf** [comment]
I'd prefer if there's an option in cython-blis to use the conda-forge `blis` package instead of vendoring in blis.

**Comment by honnibal** [comment]
@isuruf I'd like to ensure that pip and conda versions of spaCy execute the same. I'd also like to make sure that we're using the single-threaded version of Blis, which is easy to do if we vendor the the dependency.

**Comment by honnibal** [comment]
After compiling, it can't find `clock_gettime`. So I guess I need to link against `-lrt`? Is there any obvious reason why this is different? Just wondering whether there's a deeper issue here.

**Comment by honnibal** [comment]
There's a known issue with Windows Python 2.7. Currently we're advertising non-support on that. We get the following link error. I guess we're missing whatever the Windows equivalent of `-lm` is:

```
00:35:03]   bli_thread.o : error LNK2019: unresolved external symbol round referenced in function bli_thread_range_width_l
[00:35:03]   bli_util_unb_var1.o : error LNK2019: unresolved external symbol feclearexcept referenced in function bli_snormfv_unb_var1
[00:35:03]   bli_util_unb_var1.o : error LNK2019: unresolved external symbol fetestexcept referenced in function bli_snormfv_unb_var1
[00:35:03]   bli_util_unb_var1.o : error LNK2019: unresolved external symbol exp2 referenced in function bli_srandnv_unb_var1
[00:35:03]   build\lib.win-amd64-2.7\blis\cy.pyd : fatal error LNK1120: 4 unresolved externals
```

Would be nice to fix this, but otherwise we're happy to drop support for Windows on Python2.7

**Comment by honnibal** [comment]
@chrisburr I think we're finally good here 🎉

**Comment by isuruf** [comment]
>  I'd like to ensure that pip and conda versions of spaCy execute the same.

From the perspective of conda-forge, we want blis and cython-blis to do the same. That's why we don't want to vendor packages. cc @conda-forge/core

**Comment by honnibal** [comment]
@isuruf Can I name this package something different then? I understand that it shouldn't confuse people.

File: recipes/blis/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  license_file: LICENSE
+  summary: 'Fast matrix-multiplication as a self-contained Python library – no system dependencies!'
+
+  doc_url: https://github.com/explosion/cython-blis
+  dev_url: https://github.com/explosion/cython-blis
+
+extra:
+  recipe-maintainers:
+    - honnibal
+    - ines
```
Comment (scopatz) [inline recipes/blis/meta.yaml]: @ines, are you OK with being a maintainer here?

**Comment by isuruf** [comment]
@honnibal, I see that you have numpy headers vendored in. Are they from a version before numpy 1.15 or after?

File: recipes/blis/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,51 @@
+{% set name = "spacy-blis" %}
```
Comment (scopatz) [inline recipes/blis/meta.yaml]: This should probably be called `cython-blis`, given that is what the authors called it.

Comment (scopatz) [reply recipes/blis/meta.yaml]: Also, as a matter of hygiene, the directory name should match the package name,

File: recipes/blis/meta.yaml (hunk 2 above)
Comment (ines) [reply recipes/blis/meta.yaml]: @scopatz Yes 👍

File: recipes/blis/meta.yaml (hunk 3 above)
Comment (honnibal) [reply recipes/blis/meta.yaml]: Done for both.

**Comment by honnibal** [comment]
@isuruf I think after 1.15. Is there an easy way to double check?

File: recipes/cython-blis/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  sha256: {{ sha256sum }}
+
+build:
+  number: 0
+  skip: true  # [win and py27]
+
+requirements:
+  build:
+    - {{ compiler('c') }}
+    - clangdev  # [win or osx]
```
Comment (scopatz) [inline recipes/cython-blis/meta.yaml]: This doesn't seem right. Doesn't compiler('c') on osx resolve to clang?

Comment (honnibal) [reply recipes/cython-blis/meta.yaml]: I'm not sure. I'll remove it and check.

**Comment by honnibal** [comment]
@scopatz @isuruf Can this be merged please? It's starting to really affect us that this has been held up more than two weeks. @ines and I develop spaCy full-time, along with our commercial product [Prodigy](https://prodi.gy). We really need to get spaCy 2.1 out to move forward.

The release notes for spaCy v2.1: https://github.com/explosion/spaCy/releases/tag/v2.1.0a13 . We have tens of thousands of users waiting for this release.

**Comment by isuruf** [comment]
We need an answer to https://github.com/conda-forge/staged-recipes/pull/7855#issuecomment-472087335 before this can be merged.

**Comment by honnibal** [comment]
Please see: https://github.com/conda-forge/staged-recipes/pull/7855#issuecomment-472174251

> @isuruf I think after 1.15. Is there an easy way to double check?

**Comment by honnibal** [comment]
> 
https://github.com/conda-forge/staged-recipes/pull/7855#issuecomment-473192250

> 
> @isuruf I think after 1.15. Is there an easy way to double check?

**Comment by isuruf** [comment]
If it's after 1.15, the run constraint `numpy >=1.15` is wrong.

**Comment by honnibal** [comment]
Could you suggest the change? In spaCy I have: ` - {{ pin_compatible('numpy', min_pin='1.15.x') }}`. Does that look right?

**Comment by isuruf** [comment]
It depends on what the version of the headers are. If they are from numpy 1.16.2, it should be,
```
- numpy >=1.16.2
```

Normally the version of the header is the version of numpy in `host`, but since you are vendoring in numpy headers, this is not the case. Why are you vendoring numpy headers btw?

**Comment by honnibal** [comment]
I need them in the setup.py `include_dirs`. The `setup_requires` mechanism doesn't work reliably, so the best solution I've found is to vendor the headers.

**Comment by isuruf** [comment]
Following works reliably,
https://github.com/pyamg/pyamg/blob/master/setup.py#L198-L205

**Comment by honnibal** [comment]
I'm fairly sure I tried that, although possibly I did something wrong. Do you know what versions of setuptools and pip that's been tested with? I'm worried about breaking some installs.

**Comment by isuruf** [comment]
I'm not sure. Ask on numpy mailing list?

**Comment by honnibal** [comment]
Could you explain the problem here? The recipe builds, and we've been using the same solution in spaCy and this hasn't been causing any problems.

**Comment by isuruf** [comment]
You need to have a numpy version that is newer than what it was built with. You are only requiring `numpy >=1.15`, but if the headers were from `1.16`, you are going to have incompatibilities.

**Comment by honnibal** [comment]
Okay so I committed the headers November 25 2018, and numpy 1.16.0 was released January 14. So, they must be pre-1.16.0.

**Comment by honnibal** [comment]
@isuruf The build fails with:

```
ImportError: /home/conda/feedstock_root/build_artifacts/cython-blis_1552649337966/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/python2.7/site-packages/blis/cy.so: undefined symbol: clock_gettime
```

This was the error that motivated the addition of the `-lrt` arguments. Your change looks okay to me but it seems somehow that last arg isn't being read? Maybe --- not sure.

**Comment by honnibal** [comment]
@isuruf Looks good!

**Comment by honnibal** [comment]
@isuruf @scopatz If this is ready to merge, it'd be great if that could go ahead. We're going to be pushing forward with the v2.1 release today because we can't delay any longer. It would be fantastic if we could have the new version up on `conda-forge` as well.

**Comment by scopatz** [comment]
Seems OK to me.

**Comment by ines** [comment]
@scopatz This is a Travis error, right? https://travis-ci.org/conda-forge/staged-recipes/builds/507903703 Does this matter? And if so, do we need to restart the build or do anything in here?

**Comment by scopatz** [comment]
Yeah, that seems like a rate limit issue.  I have restarted

**Comment by scopatz** [comment]
Actually, no, this seems like a deeper issue.

**Pull request merged**